### PR TITLE
Module 7

### PR DESCRIPTION
**Pull request by Nastya-Korsak**
Module 7, Reflection

### CODE REVIEW

File: Module_7/Reflection/Task1/Container.cs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                .ToList();
+
+            var typesWithExportAttribute = types
+                .Single(t => t.Key == "ExportAttribute")
+                .GroupBy(t => ((ExportAttribute)t.GetCustomAttribute(typeof(ExportAttribute))).Contract == null);
+
+            typesWithExportAttribute
+                .Single(t => t.Key == true)
+                .ToList()
+                .ForEach(t => AddType(t));
```
Comment (alexander-trishin) [inline Module_7/Reflection/Task1/Container.cs:42]: Please, use `foreach` keyword. `ForEach` method is obsolete (comes from dotnet when there was no `IEnumerable`)

File: Module_7/Reflection/Task1/Container.cs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+                .GroupBy(type => CheckTypeOnExportAttribute(type) ? "ExportAttribute" :
+                (CheckTypeOnImportConstructorAttribute(type) || CheckTypeOnImportAttribute(type)) ? "ImportAttribute" : string.Empty)
+                .ToList();
+
+            var typesWithExportAttribute = types
+                .Single(t => t.Key == "ExportAttribute")
+                .GroupBy(t => ((ExportAttribute)t.GetCustomAttribute(typeof(ExportAttribute))).Contract == null);
+
+            typesWithExportAttribute
+                .Single(t => t.Key == true)
```
Comment (alexander-trishin) [inline Module_7/Reflection/Task1/Container.cs:40]: Please, use `SingleOrDefault` or `FirstOrDefault`

File: Module_7/Reflection/Task1/Container.cs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        public void AddAssembly(Assembly assembly)
+        {
+            if (assembly is null)
+            {
+                throw new ArgumentNullException(nameof(assembly));
+            }
+
+            if (_assemblies.Contains(assembly))
+            {
+                throw new DuplicateWaitObjectException(nameof(assembly), "This assembly is already added");
```
Comment (alexander-trishin) [inline Module_7/Reflection/Task1/Container.cs:24]:
```suggestion
                throw new ArgumentException(nameof(assembly), "This assembly is already added");
```

**Review by alexander-trishin — CHANGES_REQUESTED** [review]
Hi Nastya,

Good work, we're almost good to go :)

File: Module_7/Reflection/Task1/Container.cs (hunk 3 above)
Comment (Nastya-Korsak) [reply Module_7/Reflection/Task1/Container.cs]: fixed

File: Module_7/Reflection/Task1/Container.cs (hunk 2 above)
Comment (Nastya-Korsak) [reply Module_7/Reflection/Task1/Container.cs]: I added an if condition to check that the typesWithExportAttribute contains the required key, so now I think it is unnecessary

File: Module_7/Reflection/Task1/Container.cs (hunk 1 above)
Comment (Nastya-Korsak) [reply Module_7/Reflection/Task1/Container.cs]: fixed

**Review by alexander-trishin — APPROVED** [review]

**Pull request merged**